File: lg1196.cpp
#include <vector>
#include <algorithm>
#include <cstdio>
#include <numeric>

using namespace std;

vector<int> fa(30007);
vector<int> sz=fa, rk=fa;

int find(int x) {
    if (fa[x]==x) return x;
    int f = fa[x];
    fa[x] = find(fa[x]);
    rk[x] += rk[f];
    return fa[x];
}

int main() {
    fill(sz.begin(), sz.end(), 1);
    iota(fa.begin(), fa.end(), 0);
    int t;
    scanf("%d\n", &t);
    char c;
    int x, y;
    while (t--) {
        scanf("%c %d %d\n", &c, &x, &y);
        int fx=find(x), fy=find(y);
        if (c=='M') {
            fa[fx] = fy;
            rk[fx] = sz[fy];
            sz[fy] += sz[fx];
        } else {
            if (fx!=fy) {
                printf("-1\n");
                continue;
            } else {
                printf("%d\n", abs(rk[x]-rk[y])-1);
            }
        }
    }
    return 0;
}

File: lg1196.java
import java.util.*;

public class lg1196 {

    static int[] fa = new int[30007];
    static int[] sz = new int[30007];
    static int[] rk = new int[30007];

    static int find(int x) {
        if (fa[x]==x) return x;
        int f = fa[x];
        fa[x] = find(fa[x]);
        rk[x] += rk[f];
        return fa[x];
    }

    public static void main(String[] args) {
        Scanner in = new Scanner(System.in);
        Arrays.fill(sz, 1);
        for (int i=1; i<=30000; ++i) fa[i] = i;
        
        int t;
        t = in.nextInt();

        String c;
        int x, y;
        while (t-->0) {
            c = in.next();
            x = in.nextInt();
            y = in.nextInt();
            int fx = find(x), fy = find(y);
            if (c.compareTo("M")==0) {
                fa[fx] = fy;
                rk[fx] = sz[fy];
                sz[fy] += sz[fx];
            } else {
                if (fx!=fy) {
                    System.out.println("-1");
                    continue;
                } else {
                    System.out.println(Math.abs(rk[x]-rk[y])-1);
                }
            }
        }
    }
}